### DISCUSSION TITLE
Parallel execution of a panel's Prometheus queries

### DISCUSSION THREAD

**Discussion by RNabel** [grafana/grafana#54366] [Data Sources]
The Prometheus datasource executes a panel's queries serially rather than in parallel (see e.g. this [loop](https://github.com/grafana/grafana/blob/fe062f2eaac631f8405dbde6d16246d364d1e0c2/pkg/tsdb/prometheus/buffered/time_series_query.go#L111)) - is this a conscious design decision or could this be parallelized?

In my use-case, I have a panel with low 10s of queries, some quite heavy (execution time >500ms), and a backend that can handle executing them in parallel. There is one workaround: merge the queries in PromQL using `or` with `label_replace`, but this is also not very performant and highly dependent on the backend implementation of the `or` operator.

**Comment by bilbof** [comment]
When a dashboard panel has multiple Prometheus queries, these are indeed executed sequentially, as seen in the traces exported by Grafana:

![image](https://github.com/grafana/grafana/assets/8124374/197ff0a7-d54e-43b9-a075-130915fba8c9)

It would be awesome if those were executed in parallel 😄

**Comment by gtk-grafana** [comment]
I believe this was a deliberate decision made to keep CPU usage in the datasource stable to support concurrent usage. While your suggestion could potentially reduce the time it takes for the backend to return results to the frontend in some use cases, it would cause a sudden spike in CPU usage, which can crash your prometheus instance, and is not scaleable as you add more users/queries/panels/dashboards.

@itsmylife can keep me honest, do you have more context on this?

**Comment by bilbof** [comment]
It would be great to see this as an option for the user. We run reasonably large Mimir clusters which can handle + rate limit extra concurrent requests from our tenants, so we would definitely turn this on.

**Reply by gtk-grafana** [reply]
Good point, but this will increase memory usage in Grafana as well, as it must parse the responses, and store them in memory until they are ready to send to the frontend. That said I see value in initially addressing this as an experimental feature flag (i.e. if you break your Grafana by cranking up the concurrency, that's on you) for situations like this.

@itsmylife worked on the streaming parser for Prometheus though, so let's check in with him and then I can file the issue.

**Reply by itsmylife** [reply]
@gtk-grafana @bilbof I think this is not only related to the streaming parser. Frontend should send the requests parallel and on the backend-side we should run them in a goroutine. (This is a rough idea and might be completely wrong). We need to decide which places should run parallel. Like the panels with N amount of queries should run parallel. Also, there is a time interval. Where should we put the line? A good point from @gtk-grafana is it might be CPU/memory heavy for some cases. But in the end it can be a nice improvement. We should consider a simple test implementation maybe. Or maybe first a design doc to get more feedback on this. What do you think @gtk-grafana ?

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
